sway/tree/layout.c

```diff
@@ -166,9 +185,284 @@ void container_move_to(struct sway_container *container,
+	case L_VERT:
+		return dir == MOVE_UP || dir == MOVE_DOWN;
+	default:
+		return false;
+	}
+}
+
+static enum movement_direction invert_movement(enum movement_direction dir) {
+	switch (dir) {
+	case MOVE_LEFT:
+		return MOVE_RIGHT;
+	case MOVE_RIGHT:
+		return MOVE_LEFT;
+	case MOVE_UP:
+		return MOVE_DOWN;
+	case MOVE_DOWN:
+		return MOVE_UP;
+	default:
+		sway_assert(0, "This function expects left|right|up|down");
+		return MOVE_LEFT;
+	}
+}
+
+static int move_offs(enum movement_direction move_dir) {
+	return move_dir == MOVE_LEFT || move_dir == MOVE_UP ? -1 : 1;
+}
+
+/* Gets the index of the most extreme member based on the movement offset */
+static int container_limit(struct sway_container *container,
+		enum movement_direction move_dir) {
+	if (container->children->length == 0) {
+		return 0;
+	}
+	return move_offs(move_dir) < 0 ? 0 : container->children->length - 1;
+}
+
+static void workspace_rejigger(struct sway_container *ws,
+		struct sway_container *child, enum movement_direction move_dir) {
+	struct sway_container *original_parent = child->parent;
+	struct sway_container *new_parent =
+		container_split(ws, ws->layout);
+
+	container_remove_child(child);
+	for (int i = 0; i < ws->children->length; ++i) {
+		struct sway_container *_child = ws->children->items[i];
+		container_move_to(new_parent, _child);
+	}
+
+	int index = move_offs(move_dir);
+	container_insert_child(ws, child, index < 0 ? 0 : 1);
+	container_set_layout(ws,
+			move_dir == MOVE_LEFT || move_dir == MOVE_RIGHT ? L_HORIZ : L_VERT);
+
+	container_flatten(ws);
+	container_reap_empty_recursive(original_parent);
+	wl_signal_emit(&child->events.reparent, original_parent);
+	arrange_windows(ws, -1, -1);
+}
+
 void container_move(struct sway_container *container,
-		enum movement_direction dir, int move_amt) {
-	// TODO
+		enum movement_direction move_dir, int move_amt) {
+	if (!sway_assert(
+				container->type != C_CONTAINER || container->type != C_VIEW,
+				"Can only move containers and views")) {
+		return;
+	}
+	int offs = move_offs(move_dir);
+
+	struct sway_container *sibling = NULL;
+	struct sway_container *current = container;
+	struct sway_container *parent = current->parent;
+
+	if (parent != container_flatten(parent)) {
+		// Special case: we were the last one in this container, so flatten it
+		// and leave
+		update_debug_tree();
+		return;
+	}
+
+	while (!sibling) {
+		if (current->type == C_ROOT) {
+			return;
+		}
+
+		parent = current->parent;
+		wlr_log(L_DEBUG, "Visiting %p %s '%s'", current,
+				container_type_to_str(current->type), current->name);
+
+		int index = index_child(current);
+		int limit = container_limit(parent, move_dir);
+
+		switch (current->type) {
+		case C_OUTPUT: {
+			enum wlr_direction wlr_dir;
+			sway_dir_to_wlr(move_dir, &wlr_dir);
+			double ref_x = current->x + current->width / 2;
+			double ref_y = current->y + current->height / 2;
+			ref_x += current->sway_output->wlr_output->lx;
```

Comment:
Prefer to use `current->sway_output->swayc->x` (in the future, we might want to send a fake position to clients via `wlr_output`).